c++/triqs/atom_diag/impl/atom_diag.cpp

```diff
@@ -107,17 +108,48 @@ namespace triqs {
     auto atom_diag<Complex>::get_matrix_element_of_monomial(operators::monomial_t const &op_vec, int B) const
       -> std::pair<int, matrix_t> {
 
-      auto m = nda::eye<scalar_t>(get_subspace_dim(B));
+      imperative_operator<class hilbert_space, scalar_t, false>
+        monomial_op(many_body_op_t(1.0, op_vec), fops);
 
-      auto const &fops = get_fops();
-      for (int i = op_vec.size() - 1; i >= 0; --i) {
-        int ind = fops[op_vec[i].indices];
-        int Bp  = (op_vec[i].dagger ? cdag_connection(ind, B) : c_connection(ind, B));
-        if (Bp == -1) return {-1, std::move(m)};
-        m = (op_vec[i].dagger ? cdag_matrix(ind, B) : c_matrix(ind, B)) * m;
-        B = Bp;
+      state<class hilbert_space, scalar_t, true> initial_st(full_hs);
+      state<class hilbert_space, scalar_t, true> final_st(full_hs);
+
+      matrix_t m;
+      int Bp = -1;
+      bool found_nonzero_element = false;
+
+      auto const& B_fock_states = get_fock_states(B);
+      for(fock_state_t i_index = 0; i_index < B_fock_states.size(); ++i_index) {
+        fock_state_t i = B_fock_states[i_index];
+        initial_st(i) = scalar_t(1.0);
+
+        final_st = monomial_op(initial_st);
+
+        foreach(final_st, [&](fock_state_t j, scalar_t x) {
+          for(auto const& sp : sub_hilbert_spaces) {
+            if(sp.has_state(j)) {
+              Bp = sp.get_index();
+
+              if(not found_nonzero_element) {
+                m.resize(get_subspace_dim(Bp), get_subspace_dim(B));
+                m() = scalar_t(.0);
+                found_nonzero_element = true;
+              }
+
+              m(sp.get_state_index(j), i_index) = x;
+              continue;
+            }
```

Comment:
We can break out of the subspace loop as soon as we have found a single non-zero element?
There should be exactly one such element.
```suggestion
            if(sp.has_state(j) and std::abs(x) > 0.0) {
              Bp = sp.get_index();
              m = matrix_t::zeros({get_subspace_dim(Bp), get_subspace_dim(B)});
              m(sp.get_state_index(j), i_index) = x;
              break;
            }
```

Reply:
Good catch, there should have been a `break` instead of the `continue`.
There are, however, two issues with the proposed solution.

- Checking for `std::abs(x) > 0.0` is redundant, `foreach()` calls its argument only for the non-vanishing amplitudes.
- I needed `found_nonzero_element` to make sure that zeroing of `m` is done only once. Otherwise the matrix elements could get wiped out at every `i_index` iteration.

Reply:
I see, the foreach explicitly calls the `prune()`. C.f. comment below for the `found_nonzero_element`.